### ISSUE TITLE
VsCode x64 1.7 'initialize' parameters mismatch

### ISSUE THREAD

**Issue by MikhailArkhipov** [CXuesong/LanguageServer.NET#2]
Initialize request comes as

{{
  "processId": 22464,
  "capabilities": {},
  "trace": "off"
}}

while in https://github.com/CXuesong/LanguageServer.NET/blob/master/DemoLanguageServer/Services/InitializaionService.cs

`initialize(int processId, Uri rootUri, ClientCapabilities capabilities,
            JToken initializationOptions = null, string trace = null)`

expects `Uri rootUri` to always be there.

![image](https://user-images.githubusercontent.com/12820357/30617349-851f7ed4-9d4a-11e7-880d-1025f8efdae2.png)

**Comment by MikhailArkhipov** [comment]
https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#initialize does list argument as optional.

**Comment by CXuesong** [comment]
`InitializeParams.rootUri` is defined as follows
````
 	/**
	 * The rootUri of the workspace. Is null if no
	 * folder is open. If both `rootPath` and `rootUri` are set
	 * `rootUri` wins.
	 */
	rootUri: DocumentUri | null;
````
There is no question mark before the colon so I suppose this field is mandatory (except for LSP clients that relies on the deprecated `rootPath` field). From the protocol's perspective, you may assign it to `null`, but should not neglect it. 

I have tried in VSCode; however I cannot reproduce such request with missing `rootUri`, even if I have closed the opened directory. It still returns the last opened directory.

**Comment by MikhailArkhipov** [comment]
Yes, I know, it is odd. Might be a bug in VSC. `rootPath` was with `?` though. Are you using x64 VSC?

**Comment by CXuesong** [comment]
Yes… Well not exactly. I'm using v1.16.0-insiders on x64. I will try 1.16.0 now.

**Comment by CXuesong** [comment]
It works in ~~v1.16.0~~ v1.16.1. If you are testing the DemoLanguageServer, did you check out the message logs? They should be in `DemoLanguageServer\bin\Debug\netcoreapp1.1\mssages-xxxx.log`.

This is the `initialize` request received on my test run
```json
{
  "id": 0,
  "method": "initialize",
  "params": {
    "processId": 102632,
    "rootPath": "...\\MwLanguageServer\\_private",
    "rootUri": "file:///.../MwLanguageServer/_private",
    "capabilities": {
      "workspace": {
        "applyEdit": true,
        "workspaceEdit": { "documentChanges": true },
        "didChangeConfiguration": { "dynamicRegistration": false },
        "didChangeWatchedFiles": { "dynamicRegistration": false },
        "symbol": { "dynamicRegistration": true },
        "executeCommand": { "dynamicRegistration": true }
      },
      "textDocument": {
        "synchronization": {
          "dynamicRegistration": true,
          "willSave": true,
          "willSaveWaitUntil": true,
          "didSave": true
        },
        "completion": {
          "dynamicRegistration": true,
          "completionItem": { "snippetSupport": true }
        },
        "hover": { "dynamicRegistration": true },
        "signatureHelp": { "dynamicRegistration": true },
        "references": { "dynamicRegistration": true },
        "documentHighlight": { "dynamicRegistration": true },
        "documentSymbol": { "dynamicRegistration": true },
        "formatting": { "dynamicRegistration": true },
        "rangeFormatting": { "dynamicRegistration": true },
        "onTypeFormatting": { "dynamicRegistration": true },
        "definition": { "dynamicRegistration": true },
        "codeAction": { "dynamicRegistration": true },
        "codeLens": { "dynamicRegistration": true },
        "documentLink": { "dynamicRegistration": true },
        "rename": { "dynamicRegistration": true }
      }
    },
    "trace": "off"
  },
  "jsonrpc": "2.0"
}
```
Did you remove some of the JSON content for sake of clearance in the first post? 

My test environment
![image](https://user-images.githubusercontent.com/11611687/30649236-55c0baaa-9e52-11e7-9bff-136587f8a363.png)

**Comment by MikhailArkhipov** [comment]
Interesting. I was getting 'cannot find method `initialize` with matching signature' so I compiled JsonRPC.Standard with debugging and found that it fails to find candidate since there was no rootUri in the request. There data was from debugger on parameters that came to the method binder.

**Comment by CXuesong** [comment]
I think it's just weird that `rootUri` is missing.
Perhaps you need to try the latest v1.16.1? Though I suspect it won't solve the problem.
Btw I'm using vscode-languageclient@3.2.1, what's your language client version?

**Comment by MikhailArkhipov** [comment]
Looks like `2.6` - whatever VSC generated... I guess I may need new one.

        "raw": "vscode-languageclient@^2.5.0",
        "scope": null,
        "escapedName": "vscode-languageclient",
        "name": "vscode-languageclient",
        "rawSpec": "^2.5.0",
        "spec": ">=2.5.0 <3.0.0",
        "type": "range"
      },
      "C:\\RTVS\\src\\VsCode\\Extension"
    ]
  ],
  "_from": "vscode-languageclient@>=2.5.0 <3.0.0",
  "_id": "vscode-languageclient@2.6.3",
```

**Comment by CXuesong** [comment]
This is from `package.json` of my demo LSP client
```
    "devDependencies": {
        "typescript": "^2.0.3",
        "vscode": "^1.0.0",
        "mocha": "^2.3.3",
        "@types/node": "^6.0.40",
        "@types/mocha": "^2.2.32"
    },
    "dependencies": {
        "vscode-languageclient": "^3.1.0"
    }
```
Yes the versions are higher than yours.

I upgraded the versions of `vscode-languageclient` and `vscode` to the latest just now (v3.4.2 and v1,1,5); then the LSP client won't compile. Perhaps I mis-configured something. Anyway, I will try it tomorrow 🌚

The demo LSP client code currently in the repos is copied and modified several months ago from an example provided by MS. I tried to find the code again, just now, but it seems that they have changed the repos structure and I failed to find it. Anyway, you might consider starting your LSP client from the latest example provided by MS [here](https://github.com/Microsoft/vscode-extension-samples/tree/master/lsp-sample).
